File: LeetCode/Hard/LargestRectHistogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        stack = []
        ans = 0
        for i in range(len(heights)):
            start = i
            while stack and stack[-1][1] > heights[i]:
                index, height = stack.pop()
                width = i - index
                ans = max(ans, width * height)
                start = index
            stack.append((start, heights[i]))
        n = len(heights)
        while stack:
            index, height = stack.pop()
            width = n - index
            ans = max(ans, width * height)

        return ans

File: LeetCode/Hard/LargestRectHistogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        heights.push_back(0);
        stack<int> st;
        int ans = 0;
        for (int i = 0; i < heights.size(); i++) {
            while (!st.empty() && heights[i] < heights[st.top()]) {       
                int index = st.top();
                st.pop();
                int width = st.empty() ? i : i - (st.top() + 1);
                ans = max(ans, width * heights[index]);
            }
            st.push(i);
            
        }

        return ans;
    }
};